core/engine/src/vm/mod.rs

```diff
@@ -76,7 +83,7 @@ pub struct Vm {
     pub(crate) realm: Realm,
 
     #[cfg(feature = "trace")]
-    pub(crate) trace: bool,
+    pub(crate) trace: VmTrace,
```

Comment:
A thought that I've had over the last day or so is whether it would be worthwhile redesigning so that the `Trace` is a trait on `Vm`.

```rust
struct Vm<T: ExecutionTrace=()> {
    ...
    trace: T
}
```
The only issue with this that I haven't totally thought through is that I'm not sure we'd be able to feature flag the fields (all though we could probably still feature flag the method calls)

Reply:
I think the current design is fine. What would be the benefit of having it as a trait?

Reply:
Mostly I'm not sure if I'm totally in love with the `Vec<Box<dyn Tracer>>` on `VmTrace`. It works and improves the current functionality, but I'm mostly just thinking whether there may be a bit of a better design :confused: That being said, I'm not sure a theoretically better design out weighs the functionality here.

Reply:
Imo the `Vec<Box<dyn Tracer>>` is totally fine for this use case.

> Using a trait would allow the trace to be defined at compile time essentially.

I'm not sure if that is a benefit. The upside would be reducing dynamic dispatch, but if you enable tracing, you probably do not care about that last bit of performance.
Being able to change the tracing at runtime also seems like a nice feature right?

Reply:
It is a nice feature! And you're right, if if someone enables tracing then they are probably not concerned about performance. But if it's switched to a trait that doesn't mean that there couldn't be an implementation of the trait that can be updated at runtime. And if we were to update this to something without dynamic dispatching, then we even provide an impl of the trait that could be updated at runtime.

The primary difference would be that it would be a user deciding to use dynamic dispatching over Boa locking the user into using it.

Reply:
I would say the current implementation is good engough. If there are shortcomings we could change it. I would prefer that to adding a generic to the `Vm` struct. But if you think it is significantly better, go for it :)

Reply:
I do think we could offer "combinators" for tracers though? Instead of having `Vec<Box<dyn Tracer>>`, why not something like `Box<dyn Tracer>`, where we `impl<T: Tracer, U: Tracer> Tracer for (T, U)`?

Reply:
:confused: Maybe. 

There's probably a better option with how `Tracer` is defined too. I was playing around with a new design and have something like:

```rust

pub trait Tracer: std::fmt::Debug {
    fn trace_instruction(&self, ctx: &TraceDecisionCtx<'_>) -> InstructionTraceAction;
    fn trace_bytecode(&self, ctx: &TraceDecisionCtx<'_>) -> BytecodeTraceAction;
    fn emit(&self, event: &TraceEvent<'_>);
}
```

Where `TraceEvent` is something like:

```rust
pub enum TraceEvent<'a> {
    NewCallFrame(&'a NewFrameEvent<'a>),
    ExitCallFrame(&'a ExitFrameEvent<'a>),
    Instruction(&'a InstructionTrace),
    ByteCode(&'a BytecodeTraceEvent),
}
```

But that design still has `VmTrace` using a `Vec`.